### PR TITLE
Simplify TLS key creation on Windows

### PR DESCRIPTION
**Pull request by joboet**
By using a static array instead of a linked list for the storage of destructor function, the creation of new keys does not need to be synchronized. Also, running destructors should be a bit faster because of cache locality (using an atomic counter to store the maximum index of destructors avoids having to iterate over the whole array, assuming keys are mostly dense). This optimization comes at the cost of a minor runtime memory increase, however.

Using a fixed-size array is possible since [the total number of TLS keys is capped at 1088](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-local-storage). Since [TLS keys are always numbers less than 1088](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-tlssetvalue), we can just use the value of the key to index into the array.

@rustbot label +O-windows-gnu

### CODE REVIEW

File: library/std/src/sys/windows/thread_local_key.rs [hunk 1]
```diff
@@ -19,106 +21,92 @@ pub type Dtor = unsafe extern "C" fn(*mut u8);
...
-//   means that we have a global list of destructors for each TLS key that
-//   we know about.
-// * When a thread exits, we run over the entire list and run dtors for all
-//   non-null keys. This attempts to match Unix semantics in this regard.
-//
-// This ends up having the overhead of using a global list, having some
-// locks here and there, and in general just adding some more code bloat. We
-// attempt to optimize runtime by forgetting keys that don't have
-// destructors, but this only gets us so far.
+// Since the maximum number of keys is 1088 and key values are always lower than
```
Comment (fbstj) [inline library/std/src/sys/windows/thread_local_key.rs:24]: might be worth adding a link to cite this?

File: library/std/src/sys/windows/thread_local_key.rs [hunk 2]
```diff
@@ -19,106 +21,93 @@ pub type Dtor = unsafe extern "C" fn(*mut u8);
...
-    if let Some(f) = dtor {
-        register_dtor(key, f);
+
+    if let Some(dtor) = dtor {
+        DTORS[key as usize].store(mem::transmute::<Dtor, *mut ()>(dtor), Relaxed);
+        HIGHEST.fetch_max(key as usize, Relaxed);
+        // If the destructors are run in a signal handler running after this
+        // code, we need to guarantee that the changes have been performed
+        // before the handler is triggered.
+        compiler_fence(Release);
```
Comment (thomcc) [inline library/std/src/sys/windows/thread_local_key.rs:53]: This micro-optimization doesn't really feel likely to be worth it, and it's plausibly not even totally sound.

We should just use the correct orderings for these atomics -- it really shouldn't make a performance difference and is easier to reason about.

**Review by thomcc — COMMENTED** [review]
I need to think a bit about this. In general I'm not sure it's worth hard-coding limits like this, even if they can't change (at least not on existing targets).

CC @ChrisDenton (we've chatted at length about the TLS situation on Windows -- thoughts here?)

**Comment by bjorn3** [comment]
Shouldn't we rather work around the 1088 cap on number of tls keys instead of fixating on it even more? Libstd consumes a fair amount of tls keys by itself (a hello world seems to use 20 keys, might have counted wrong though). I can easily imagine running out of tls keys in complex programs in the future. For example once more libraries start using rust internally and include their own copy of libstd.

**Comment by thomcc** [comment]
> Shouldn't we rather work around the 1088 cap on number of tls keys instead of fixating on it even more? Libstd consumes a fair amount of tls keys by itself (a hello world seems to use 20 keys, might have counted wrong though). I can easily imagine running out of tls keys in complex programs in the future. For example once more libraries start using rust internally and include their own copy of libstd.

20 is surprising. I went through looking for these recently, and didn't find nearly that many. Looking now, I only count 6 that arent part of tests:
1. Hashmap keys
2. Thread info
3. thread panic state (not used until a thread panics).
4. stdio output capture (only used by libtest).
5. threadid address byte in recursive mutex.
6. one in library/backtrace, probably hard to combine with anything else without coupling (altho it should use a const ctor).

3 and 4 don't use slots in most programs. 5 could be avoided (at the cost of either perf or system specific code), and 1/2 can be combined (at the cost of coupling).

There are also a bunch in library/proc_macro, but I think they don't matter?

> For example once more libraries start using rust internally and include their own copy of libstd

I'm not sure it's generally okay to have multiple copies of libstd in your program at the same time. I guess it's probably fine if they're all fully independent.

---

All that said, I think in practice we hope to eventually have`#[cfg(target_thread_local)]` enabled for most/all windows targets (disabled by https://github.com/rust-lang/rust/issues/95429, which might be fixable?), so the hard-coded limit might not matter that much here.

(Although... IIRC a limit like this might still exist for static TLS on windows, so maybe it's still worth thinking about)

**Comment by ChrisDenton** [comment]
> (Although... IIRC a limit like this might still exist for static TLS on windows, so maybe it's still worth thinking about)

Static TLS is just one big array embedded in the exe. I guess the only limit would be the maximum size of a single section but I've never needed to find out.

**Comment by thomcc** [comment]
> In contrast, [TlsAlloc](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-tlsalloc) has this to say:
>
> > The value of the TLS index should be treated as an opaque value; do not assume that it is an index into a zero-based array.

Ah, nice catch. IMO we definitely shouldn't do this then.

**Comment by bjorn3** [comment]
> I'm not sure it's generally okay to have multiple copies of libstd in your program at the same time. I guess it's probably fine if they're all fully independent.

Indeed. Linking multiple rust staticlib together is not fine as the libsyd symbols are exported outside of the staticlib. Linking multiple cdylib together is fine as it only exports user defined `#[no_mangle]` symbols.

> 20 is surprising. I went through looking for these recently, and didn't find nearly that many.

I counted the amount of callers of `thread_local_key::StaticKey::lazy_init` as proxy, but that may indeed have been wrong.

**Comment by joboet** [comment]
Ok, so this is definitely not possible then. An alternative solution would be to use fiber local storage, as it [allows destructors](https://learn.microsoft.com/en-us/windows/win32/api/fibersapi/nf-fibersapi-flsalloc) and can be called even from outside fiber environments ([TLS was implemented on top of FLS](https://github.com/llvm/llvm-project/blob/e030be64d8c43a56b60a90b70765fc795e177e9c/libcxx/src/support/win32/thread_win32.cpp#L251-L272) on older Windows versions, so this is probably a stable guarantee). This is actually the solution [used by libc++](https://github.com/llvm/llvm-project/blob/e030be64d8c43a56b60a90b70765fc795e177e9c/libcxx/src/support/win32/thread_win32.cpp#L251-L272), so we would be in good company. However, it would be quite surprising for projects using native Windows fibers as a green thread implementation, assuming those exist (I could not find any).

**Comment by thomcc** [comment]
I'm pretty sure there are projects using fibers as green threads, and even if there aren't IMO we should support code that wants to do that (at least, I think this optimization is not worthwhile enough to make a difference there). So, I'm not sure `std` is the right place for that. ISTM there's just a lot of ways it could go wrong (issues around `Thread`/`ThreadId`, the fact that the dtor runs on fiber destruction instead of thread destruction, ...).

I have done it in my own code before though -- it's nice to have a version of TlsAlloc that takes a dtor, it's just only really a good idea if you won't be using fibers for anything else (or know you're using it for a case where it wouldn't matter).

**Pull request closed without merging**